### PR TITLE
stack based splitting of args

### PR DESCRIPTION
**Pull request by P4N74**
r_str_word_set0_stack. (works using stack taking care of nested paranthesis, brackets and braces). The test case might fail since it does not require `"` for nested things which it did earlier.

### CODE REVIEW

File: libr/util/str.c [hunk 1]
```diff
@@ -372,6 +373,77 @@ R_API int r_str_word_set0(char *str) {
...
+		return 0;
+	}
+	for (i=0; str[i] && str[i+1]; i++) {
+		if (i > 0 && str[i-1] == ' ' && str[i] == ' ') {
+			int len = strlen (str+i)+1;
+			memmove (str+i, str+i+1, len);
+			i--;
+		}
+		if (i == 0 && str[i] == ' ') {
+			memmove (str+i, str+i+1, strlen(str+i)+1);
```
Comment (radare) [inline libr/util/str.c]: spaces around + and before (

File: libr/util/str.c [hunk 2]
```diff
@@ -372,6 +373,77 @@ R_API int r_str_word_set0(char *str) {
...
+					(*(char *)pop == '{' && *p == '}') ||
+					(*(char *)pop == '[' && *p == ']')) {
+					continue;
+				}
+			}
+		}
+
+		if (*p == ' ') {
+			if (p > str && !*q) {
+				memmove (p, p+1, strlen (p+1)+1);
```
Comment (radare) [inline libr/util/str.c]: spaces around +

File: libr/util/str.c [hunk 3]
```diff
@@ -372,6 +373,77 @@ R_API int r_str_word_set0(char *str) {
...
+				if (*q == '\\') {
+					*q = ' ';
+					continue;
+				}
+				p--;
+			}
+			if (r_stack_is_empty(s)) {
```
Comment (radare) [inline libr/util/str.c]: space before (

File: libr/util/str.c [hunk 4]
```diff
@@ -372,6 +373,77 @@ R_API int r_str_word_set0(char *str) {
...
+	s = r_stack_new (5); //Some random number
+	for (i = 1, p = str; *p; p++) {
+		q = p - 1;
+		if (p > str && (*q == '\\')) {
+			memmove (q, p, strlen (p) + 1);
+			p--;
+			continue;
+		}
+		if (*p == '(' || *p == '{' || *p == '[') {
+			r_stack_push(s, (void *)p);
```
Comment (radare) [inline libr/util/str.c]: missing space before (

File: libr/util/str.c [hunk 5]
```diff
@@ -372,6 +373,77 @@ R_API int r_str_word_set0(char *str) {
...
+			continue;
```
Comment (radare) [inline libr/util/str.c]: dont else if the block breaks/continue the loop  to make that clear

File: libr/util/str.c [hunk 6]
```diff
@@ -372,6 +373,77 @@ R_API int r_str_word_set0(char *str) {
...
+		} else if (*p == '\'' || *p == '"') {
+			pop = r_stack_pop (s);
+			if (pop && *(char *)pop != *p) {
+				r_stack_push (s, pop);
+				r_stack_push (s, (void *)p);
+			} else if (!pop) {
+				r_stack_push (s, (void *)p);
+			}
+			continue;
+		} else if (*p == ')' || *p == '}' || *p == ']') {
```
Comment (radare) [inline libr/util/str.c]: same here, no need for an else block

File: libr/util/str.c [hunk 7]
```diff
@@ -372,6 +373,77 @@ R_API int r_str_word_set0(char *str) {
...
+	}
+	s = r_stack_new (5); //Some random number
+	for (i = 1, p = str; *p; p++) {
+		q = p - 1;
+		if (p > str && (*q == '\\')) {
+			memmove (q, p, strlen (p) + 1);
+			p--;
+			continue;
+		}
+		if (*p == '(' || *p == '{' || *p == '[') {
```
Comment (radare) [inline libr/util/str.c]: what about using a switch case? instead of 9 ifs

Comment (P4N74) [inline libr/util/str.c]: I didn't realize I had that many if's. I will add switch for this if statement.

**Comment by P4N74** [comment]
@radare I was about to update the pr with fixing all the comments :( I was just doing some test before I was pushing.

**Comment by radare** [comment]
oops. send another one. :)

**Pull request merged**